### PR TITLE
[ROCm][Bugfix] Densify strided activations for wvSplitKQ

### PR DESCRIPTION
**Pull request by JohnQinAMD**
## Purpose

Prevent ROCm skinny GEMMs from reading past a packed activation view.
`wvSplitK` and `wvSplitKrc` do not consume strides, so their dispatch now
materializes the activation only when its 2-D layout is not exactly dense
row-major.

Kimi-K3 KDA projects several outputs into one row and passes the 128-element
`f_a` slice to `f_b_proj`. Under TP8 that view has `stride=(6288, 1)` and
`storage_offset=6144`. The skinny kernel preloads a flat span; the view is valid
for PyTorch indexing but the preload extends 12,288 bytes past it.

| Rows | `is_contiguous()` | Available | Flat preload | Over-read |
| ---: | ---: | ---: | ---: | ---: |
| 1 | `True` | 288 B | 12,576 B | 12,288 B |
| 4 | `False` | 38,016 B | 50,304 B | 12,288 B |

The singleton case is why `is_contiguous()` is insufficient: PyTorch ignores
singleton-dimension strides. Dense inputs, fallback GEMMs, NVIDIA, HIP kernel
code, and dispatch thresholds are unchanged.

## Test plan

Tested on 8x MI355X (`gfx950`) with the public Kimi-K3 image. Models:
`moonshotai/Kimi-K3@9f62e4e9` and `Inferact/Kimi-K3-DSpark@cf6b8244`.
The command below pins the control and PR source revisions.

Fetch and verify the candidate:

```bash
git clone https://github.com/vllm-project/vllm.git vllm
git -C vllm fetch origin refs/pull/50618/head:pr-50618
test "$(git -C vllm rev-parse pr-50618)" = \
  aa780f0db6e5f467378856dffda1e939bca53604
git -C vllm checkout --detach 9a7ae4b80866a322678ece31ca2854d9a174ab8d
git -C vllm diff 9a7ae4b80866a322678ece31ca2854d9a174ab8d..pr-50618 \
  --binary | git -C vllm apply -
git -C vllm diff --check
```

Use the complete verified vLLM tree as an overlay on the public image; do not
copy individual files. Verify `import vllm, vllm._C` before testing. Run:

```bash
python -m pytest -q \
  tests/model_executor/layers/test_rocm_unquantized_gemm.py
```

Launch TP8 with both exact model revisions:

```bash
VLLM_ROCM_USE_AITER=1 VLLM_ROCM_USE_AITER_FP4BMM=1 \
VLLM_ROCM_USE_SKINNY_GEMM=1 AITER_SITUV2_A8W4=1 \
AITER_BF16_FP8_MOE_BOUND=0 VLLM_USE_BREAKABLE_CUDAGRAPH=0 \
vllm serve /model --served-model-name moonshotai/Kimi-K3 \
  --tensor-parallel-size 8 --trust-remote-code --moe-backend auto \
  --gpu-memory-utilization 0.88 --max-num-seqs 4 \
  --max-num-batched-tokens 4096 --max-model-len 1048576 \
  --enable-prefix-caching \
  --compilation-config '{"cudagraph_capture_sizes":[1,4,64]}' \
  --speculative-config \
  '{"method":"dspark","model":"/draft","num_speculative_tokens":7,"attention_backend":"TRITON_MLA"}'
```

The focused test injects packed singleton and multi-row views into both skinny
dispatches and asserts that the native kernels receive dense `(128, 1)`
activations. The real-workload layout can also be proved without issuing the
unsafe read by logging shape, stride, storage offset, remaining storage bytes,
and `stride(0) * rows * element_size()` immediately before dispatch.

## Test results

- Pinned-public focused suite: **4/4 passed in 4.46 s**.
- Native MI355X B1/B4 checks reached `wvSplitK` with dense activations and
  matched `torch.nn.functional.linear` exactly (`max_abs_error=0`).
- TP8 loaded 96/96 target shards and the draft; graph capture completed at
  1/4/64 tokens; health and generation passed with no memory-access fault.
- Full source-isolated GSM8K from the preceding runtime revision: control
  **1274/1319**, candidate **1276/1319**, zero invalid responses, 13 wins/11
  losses (`p=0.8388`). The final commit changes only the stride predicate and
  singleton test coverage; the full run was not repeated at that final head.
- Changed-file pre-commit, `git diff --check`, and DCO passed.

## Limits

The unpatched fault is allocator-layout dependent: the unsupported read faults
only when it crosses an unmapped page. The deterministic stride/storage probe
establishes the over-read even when a serving run happens not to fault.

## Tool assistance

Claude Opus and OpenAI Codex assisted with implementation, tests, log/source
analysis, and drafting this description.

---

## Update 2026-08-18 — rewritten after re-checking against current upstream

**The original bug no longer reproduces.** #50907 densifies both skinny call sites
in `layers/utils.py`, covering what this PR fixed plus the `LLMM1` path it missed.
Checked over 139 layouts, including the single-row cases a contiguity flag cannot
express: zero over-read in all of them.

**The same defect is live in `wvSplitKQ`, which is what this PR now fixes.**
Identical staging bound (`Kap * N`, `Kap = in_a.stride(0)`), and its call site
`scaled_mm/rocm.py` never densifies.

There is a reachable in-tree path to it, and it runs entirely through
`is_contiguous()` reporting True for a single-row view:

| step | what happens |
|---|---|
| `x` is a single-row activation whose row stride exceeds K | `x.is_contiguous()` is **True** — PyTorch ignores the stride of a size-1 dim |
| `QuantFP8.forward_hip` | gates the aiter path on exactly that flag (`input_quant_fp8.py:144`), so it is taken |
| `rocm_aiter_ops.per_tensor_quant` | allocates with `torch.empty_like(x, dtype=...)`, which **preserves the row stride** |
| `ScaledMMLinearKernel` | `x_data.view(-1, shape[-1])` is a no-op on an already-2-D tensor |
| `wvSplitKQ` | receives stride(0) > K and stages past the extent |

Measured: at 1 row the stride survives `empty_like` as `(6288, 1)`; at 2 and 4 rows
`is_contiguous()` is False, the aiter gate rejects it, and the output comes back
dense. So the exposure is specifically single-row decode — which is what this
dispatch is for (`A.shape[0] <= 4`).

Config required: fp8 W8A8 with **per-tensor** activation and weight scales,
`VLLM_ROCM_USE_SKINNY_GEMM` (default on), `B.shape[0] % 16 == 0`,
`B.shape[1] % 16 == 0`.

```python
# repro.py -- MI355X (gfx950), ROCm 7.2, torch 2.11.0
import torch, vllm._custom_ops as ops

K, MULT = 1024, 8192
cu = torch.cuda.get_device_properties(0).multi_processor_count
fp8 = torch.float8_e4m3fn

wide = torch.randn(1, K * MULT, dtype=torch.float32, device="cuda").to(fp8)
act = wide[:, K * (MULT - 1):]      # tail slice: valid data ends at the storage end
w = torch.randn(64, K, dtype=torch.float32, device="cuda").to(fp8)
s = torch.ones(1, dtype=torch.float32, device="cuda")

print(f"act{tuple(act.shape)} stride={act.stride()} "
      f"-> kernel walks {act.stride(0)} elems, {K} valid")
ops.wvSplitKQ(w, act, torch.bfloat16, s, s, cu, None)
torch.cuda.synchronize()
```

```console
$ PYTORCH_NO_CUDA_MEMORY_CACHING=1 python repro.py
act(1, 1024) stride=(8388608, 1) -> kernel walks 8388608 elems, 1024 valid
Memory access fault by GPU node-3 (Agent handle: 0x3ce8f1c0) on address 0x7090fb800000
```

The env var only makes the fault visible: pooled allocations leave the bytes past a
tensor mapped, so the over-read normally reads a neighbour instead of faulting.
Staging bounds by the stride, compute bounds by `K`, so results stay correct and no
`allclose` test can catch this. Shrinking the stride so the over-read is 256 /
16,128 / 1,047,552 elements instead, it does not fault at all and the results are
correct — in production that makes it an occasional crash unrelated to its cause.

### Changes

- `scaled_mm/rocm.py`: densify before `wvSplitKQ`. Tests the stride explicitly
  rather than calling `.contiguous()`, which is a no-op on a single-row view.
- `skinny_gemms.cu`: state the contract `wvSplitKQ` already depends on, so a future
  call site fails at the boundary instead of over-reading.

### Test plan / results

```bash
pytest tests/model_executor/layers/test_rocm_scaled_mm_strided.py -q
```

Six tests, none vacuous: the four strided cases and the kernel-contract case fail
without this PR (the last as `DID NOT RAISE RuntimeError`); dense passthrough passes
either way. `ruff check`/`format` clean at the pinned v0.14.0.

The kernel-contract test needs the rebuilt extension, so I built it: with the
`TORCH_CHECK` compiled in it passes, and against a build without it the same test
reports `DID NOT RAISE RuntimeError`.

Tool assistance: analysis, repro and this description produced with Claude Code; I
reviewed every changed line and ran the above on MI355X.

### CODE REVIEW

**Comment by AndreasKaratzas** [comment]
Be careful with your usage of AI. This PR especially in the validation parts tells me that it was a pure copy paste.  Reproduction also tells me nothing here. What was the command that you run, and what was the bug?

**Comment by JohnQinAMD** [comment]
@AndreasKaratzas  Thanks for calling this out. I use official  vLLM image and rewrote the PR description so it show exact the bug and rootcause.  The exact `docker run` command, image digest, model paths, diagnostic guard, and observed output are now under
**Reproduction**.

The bug is at the `wvSplitK` dispatch boundary. Kimi-K3's packed KDA projection creates `f_a` as a `(N, 128)` view with stride `(6288, 1)` and storage offset 6,144. The gfx950 skinny kernel stages `stride(0) * N` elements linearly into
LDS, so the real `N=4` input reads 12,288 bytes past its backing allocation. That is the source of the allocator-dependent GPU memory fault.

The latest head also covers an easy-to-miss `N=1` case: PyTorch reports the singleton view as contiguous despite its 6,288 row stride, so `is_contiguous()`/`.contiguous()` is insufficient. The implementation now checks for exact dense stride and clones only immediately before the two skinny kernel calls.

Current validation is 4/4 focused tests, native MI355X `wvSplitK` comparison for the real Kimi-K3 `N=1` and `N=4` layouts (`max_abs_error=0.0`), and a clean TP8 official-image startup with graph captures `[1, 4, 64]` plus a successful generation request. All local pre-commit hooks pass.

**Comment by hongxiayang** [comment]
@JohnQinAMD : can you please resolve the conflicts, and re-run with the current vllm upstream nightly image and update your test to check whether the issue still holds. Be sure to describe clearly how to reproduce the problem using up-to-date recipe, before and after. thanks

**Comment by JohnQinAMD** [comment]
@hongxiayang rebased and re-ran on current `main`: **the bug this was filed for is gone** — #50907 covers it, verified across 139 layouts with zero over-read.

But the same defect is still live in `wvSplitKQ`, whose call site never densifies, and I reproduced it as a GPU fault on MI355X. So rather than close this I have rewritten the PR to fix that: densify at the call site plus a contract check in the kernel. Three files, 131 lines, 112 of them tests; five of the six tests fail without the change.

Reachability, the repro script and its output are in the updated description.
